src/test/groovy/unit/org/pih/warehouse/requisition/RequisitionServiceSpec.groovy

```diff
@@ -0,0 +1,159 @@
+package unit.org.pih.warehouse.requisition
+
+import grails.testing.gorm.DataTest
+import grails.testing.services.ServiceUnitTest
+import org.pih.warehouse.core.IdentifierService
+import org.pih.warehouse.core.Location
+import org.pih.warehouse.core.Person
+import org.pih.warehouse.core.User
+import org.pih.warehouse.inventory.InventoryService
+import org.pih.warehouse.inventory.Transaction
+import org.pih.warehouse.picklist.Picklist
+import org.pih.warehouse.requisition.Requisition
+import org.pih.warehouse.requisition.RequisitionService
+import org.pih.warehouse.requisition.RequisitionStatus
+import org.pih.warehouse.requisition.RequisitionType
+import spock.lang.Shared
+import spock.lang.Specification
+import spock.lang.Unroll
+import grails.validation.ValidationException
+
+@Unroll
+class RequisitionServiceSpec extends Specification implements ServiceUnitTest<RequisitionService>, DataTest {
+
+    @Shared
+    InventoryService inventoryService
+
+    @Shared
+    IdentifierService identifierService
+
+    void setupSpec() {
+        mockDomain Requisition
+        mockDomain Picklist
+    }
+
+    void setup() {
+        inventoryService = Spy(InventoryService) {
+            generateTransactionNumber() >> UUID.randomUUID().toString()
+        }
+        identifierService = Stub(IdentifierService) {
+            generateRequisitionIdentifier() >> UUID.randomUUID().toString()
+        }
+        inventoryService.identifierService = identifierService
+        service.inventoryService = inventoryService
+        service.identifierService = identifierService
+    }
+
+
+    void 'saveRequisition saves the given requisition'() {
+        given:
+        Requisition requisition = new Requisition(id: 1)
+
+        when:
+        Requisition savedRequisition = service.saveRequisition(requisition)
+
+        then:
+        savedRequisition != null
+    }
+
+    void 'issueRequisition should throw an exception when picklist is missing'() {
+        given:
+        Requisition requisition = new Requisition(id: 1)
+        String comment = "Comment to requisition with errors"
+
+        and:
+        Requisition persistedRequisition = service.saveRequisition(requisition)
+
+        when:
+        service.issueRequisition(persistedRequisition, Mock(User), Mock(Person), comment)
+
+        then:
+        thrown(ValidationException)
+    }
+
+    void 'issueRequisition should throw an exception when inventory service is unable to save local transfer'() {
+        given:
+        Requisition requisition = new Requisition(id: 1)
+        String comment = "Comment to requisition with errors"
+
+        and:
+        Picklist.metaClass.static.findByRequisition = { Requisition foundRequisition -> return new Picklist(requisition: foundRequisition)
+        }
```

Comment:
nitpick: I'd either move the '}' to the above line or move the return statement down to its own line. Same comment for the next test.

It's weird that grails and gorm doesn't really provide a method of stubbing domain methods. This feels weird to have to do but I couldn't find a better solution from googling.

Reply:
> nitpick: I'd either move the '}' to the above line or move the return statement down to its own line. Same comment for the next test.

I did it like that before and the run autoformatting in Intellij :smile:

Reply:
> It's weird that grails and gorm doesn't really provide a method of stubbing domain methods.

There are. We just need to do some more work on testing to start using them. I'm hoping a test-a-thon exercise will help us uncover more testing best practices.